zerver/tests/test_audit_log.py

```diff
@@ -1151,3 +1152,27 @@ def test_change_user_group_subgroups_memberships(self) -> None:
                 orjson.loads(assert_is_not_none(audit_log_entries[i].extra_data)),
                 {"supergroup_id": user_group.id},
             )
+
+    def test_user_group_deletion(self) -> None:
+        hamlet = self.example_user("hamlet")
+        user_group = check_add_user_group(
+            hamlet.realm, "foo", [hamlet], description="Some description", acting_user=hamlet
+        )
+        now = timezone_now()
+        check_delete_user_group(user_group.id, acting_user=hamlet)
+        audit_log = RealmAuditLog.objects.filter(
+            realm=hamlet.realm,
+            event_time__gte=now,
+            event_type=RealmAuditLog.USER_GROUP_DELETED,
+        )
+        self.assert_length(audit_log, 1)
+        self.assertEqual(audit_log[0].acting_user, hamlet)
+        self.assertDictEqual(
+            orjson.loads(assert_is_not_none(audit_log[0].extra_data)),
+            {
+                "name": "foo",
+                "realm_id": hamlet.realm_id,
+                "description": "Some description",
+                "is_system_group": False,
+            },
+        )
```

Comment:
Can we have a test for what happens with nested groups?  That is, if A is in B, and I delete B, does A get deleted?  Does it get an USER_GROUP_DELETED audit log entry?  I expect "yes" and "yes," but I suspect the code as written is "yes" and "no" respectively.

Reply:
Did you mean audit log entries for `USER_GROUP_DIRECT_SUPER/SUBGROUP_MEMBERSHIP_DELETED` event? I think we do a cascade delete only for the `GroupGroupMembership` rows when a related user group is deleted, but not for the user groups themselves. I think we do need to generate these audit log entries.

Reply:
No, I mean what happens to subgroups when you delete the supergroup.

Reply:
The subgroup remains as it is when the supergroup is deleted and as mentioned above by Zixuan the `GroupGroupMembership` objects will be deleted.

Reply:
Ah -- OK.  I guess it's necessary to leave the sub-group, particularly if we're allowing subgroup C to be in both group A and group B.  Makes sense!  But we do at least need the `USER_GROUP_DIRECT_SUPER` event.